plugin/file: Fix in wrong answers returned when wildcard and concrete records exist

xdu31 opened the pull request:
### 1. Why is this pull request needed and what does it do?
It fixes in file plugin wrong answers returned when wildcard and concrete records exist:
<pre>
; example.org test file with wildcard corner cases
$TTL 3600
example.org.		IN	SOA	sns.dns.icann.org. noc.dns.icann.org. 2015082541 7200 3600 1209600 3600  
example.org.		IN	NS	b.iana-servers.net.  
*.d.example.org.        IN      A       127.0.1.53  
r.c.d.example.org.      IN      A       127.0.1.52
</pre>
As per RFC 4592, the query `something.r.c.d.example.com. , A` should not match the wildcard record `*.d.example.com. , A` as `r.c.d.example.com. , A` exists. It is considered the closest encloser, and since there is neither something domain nor a wildcard record under it, the answer has to be NXDOMAIN.
CoreDNS exhibit the same behavior even for queries like `something.c.d.example.com. , A`. The expected response is NXDOMAIN, but CoreDNS returns the wildcard record.

### 2. Which issues (if any) are related?
https://github.com/coredns/coredns/issues/4290

### 3. Which documentation changes (if any) need to be made?

### 4. Does this introduce a backward incompatible change or deprecation?
No

Commit:
plugin/file: Fix in wrong answers returned when wildcard and concrete records exist

Signed-off-by: Jason Du <xdu@infoblox.com>

miekg commented on plugin/file/lookup.go:
```diff
@@ -197,11 +197,22 @@ func (z *Zone) Lookup(ctx context.Context, state request.Request, qname string)
 	if wildElem != nil {
 		auth := ap.ns(do)
+		if foundCe && dns.CountLabel(ce.Name())+1 > dns.CountLabel(wildElem.Name()) {
+			// wildcard denial
```
this comment is a bit terse to explain the previous if statement, please expand a bit

xdu31 replied:
Hi @miekg This is the case described in https://github.com/coredns/coredns/issues/4290, when the record "something.r.c.d.example.com" is matching "*.d.example.com" wildcard, "r.c.d.example.com", its closest encloser (CE) can be found, however, its CE is not directly under the wildcard (cntLabel(CE) + 1 > cntLabel(wildcard)), the record needs to be resolved as nxdomain.

miekg commented on plugin/file/lookup.go:
```diff
@@ -232,6 +243,7 @@ func (z *Zone) Lookup(ctx context.Context, state request.Request, qname string)
 			rrs = append(rrs, sigs...)
 
 		}
+
```
superfluous change

miekg commented on plugin/file/lookup.go:
```diff
@@ -258,10 +270,8 @@ func (z *Zone) Lookup(ctx context.Context, state request.Request, qname string)
-
 		// wildcard denial only for NXDOMAIN
-		if found {
+		if foundCe {
```
is `found` actually still used?

xdu31 replied:
> is `found` actually still used?

Yes, in a few other places

https://github.com/coredns/coredns/blob/d6d8a9974bc45855a7d8a89e9e452cc9b6301d0e/plugin/file/lookup.go#L95

https://github.com/coredns/coredns/blob/d6d8a9974bc45855a7d8a89e9e452cc9b6301d0e/plugin/file/lookup.go#L162

https://github.com/coredns/coredns/blob/d6d8a9974bc45855a7d8a89e9e452cc9b6301d0e/plugin/file/lookup.go#L167

https://github.com/coredns/coredns/blob/d6d8a9974bc45855a7d8a89e9e452cc9b6301d0e/plugin/file/lookup.go#L250

Commit:
Remove superfluous change

Signed-off-by: Jason Du <xdu@infoblox.com>

Commit:
Re-implementation and new test case
If the domain's terminal is neither the matching wildcard, nor a domian directly
under the wildcard, return NXDOMAIN

Signed-off-by: Jason Du <xdu@infoblox.com>

chrisohaver commented on plugin/file/lookup.go:
```diff
@@ -239,8 +255,8 @@ func (z *Zone) Lookup(ctx context.Context, state request.Request, qname string)
 	// must be an empty-non-terminal. If so, we do the proper NXDOMAIN handling, but set the rcode to be success.
-	if x, found := tr.Next(qname); found {
-		if dns.IsSubDomain(qname, x.Name()) {
+	if nextElem != nil {
```
That this logic has essentially changed from `tr.Next(qname)` to `tr.Next(parts)`, seems like it might break empty non-terminal resolution.

chrisohaver replied:
> That this logic has essentially changed from `tr.Next(qname)` to `tr.Next(parts)`, seems like it might break empty non-terminal resolution.

I don't see an existing test that validates queries for empty non-terminals should get a 'NOERROR' response. Should add it if one doesn't currently exist. e.g. in `lookup_test.go`...

`dbMiekNL` +=
```
test.ent	IN	A	139.162.196.79
```

and `dnsTestCases` +=
```
	{
		Qname: "ent.miek.nl.", Qtype: dns.TypeA,
		Ns: []dns.RR{
			test.SOA("miek.nl.	1800	IN	SOA	linode.atoom.net. miek.miek.nl. 1282630057 14400 3600 604800 14400"),
		},
	},
```

xdu31 replied:
Hi @chrisohaver I reverted back the tr.Next(qname) and added a new test case in lookup_test.go.

chrisohaver replied:
Cool - thanks.  Looks like the other two wildcard related tests in azure and clouddns are passing now after that change too.

xdu31 replied:
Hi @chrisohaver, the idea is to iteratively find the next record that is the subdomain of the parts, it will finally find the last(longest) matching subdomain record.

For the zone file in `cornerCasesWildcard`
<pre>
     org
      |
   example
      |
      d
     / \
    c   *
    |
    r
</pre>

`example.org.`, ne is `*.d.example.org.`
`d.example.org.` ne is `*.d.example.org.`

`c.d.example.org.` ne is `r.c.d.example.org.`
`r.c.d.example.org.` ne is `r.c.d.example.org.`
...

`something.d.example.org.` ne is `*.d.example.org.` (inherent from `d.example.org.` as `*.d.example.org.` is not a parent of `something.d.example.org.`)
...

If number of labels of ne is greater than number of labels of the matching wildcard, we want to return NXDOMAINS

Commit:
Fix empty non-terminal & add test case

Signed-off-by: Jason Du <xdu@infoblox.com>

Commit:
Cleanup

Signed-off-by: Jason Du <xdu@infoblox.com>

chrisohaver commented on plugin/file/lookup.go:
```diff
@@ -201,8 +207,18 @@ func (z *Zone) Lookup(ctx context.Context, state request.Request, qname string)
 	// Found wildcard.
 	if wildElem != nil {
-		auth := ap.ns(do)
+		// if the domain's ternimal is neither the matching wildcard, nor a domain directly under the wildcard:
```
```suggestion
		// if the domain's terminal is neither the matching wildcard, nor a domain directly under the wildcard:
```

chrisohaver commented on plugin/file/lookup.go:
```diff
@@ -92,6 +92,12 @@ func (z *Zone) Lookup(ctx context.Context, state request.Request, qname string)
 			break
 		}
 
+		if nextElem, found = tr.Next(parts); found {
```
Tests fail if the following record is added to the zone: `x.&.d.example.org.        IN      A       127.0.1.53`.

I think it's because of an assumption that a `*` element will always be ordered in the tree before any other element in the same sub domain (and `&` is sorted before `*`).  
E.g. the following record does not cause tests to fail: `x.+.d.example.org.        IN      A       127.0.1.53` presumably because `+` sorts after `*`.

miekg replied:
sorting and tr.Next don't do what you think. 

and in general I think this PR just happens to implement what the tests hand it

Commit:
Fix case on domain name with character before * and add more test cases

Signed-off-by: Jason Du <xdu@infoblox.com>

Commit:
Merge branch 'master' of https://github.com/coredns/coredns into file-plugin-wildcard

chrisohaver commented on plugin/file/wildcard_test.go:
```diff
@@ -296,3 +366,13 @@ example.org.		IN	NS	b.iana-servers.net.
+example.org.		IN	NS	b.iana-servers.net.
+*.d.example.org.        IN      A       127.0.1.53
+z.+.d.example.org.		IN		A		127.0.1.53
+x.&.d.example.org.		IN		A		127.0.1.53
```
These IPs should be different from the wild card entry IP - So that an answer from these records can be discerned from an (incorrect) answer synthesized from the wildcard record.

xdu31 replied:
@chrisohaver Change made.

chrisohaver reviewed (changes requested):
Looks good.  The latest version of the logic feels more intuitive IMO. Thanks!

One minor comment about the test data, but other than that LGTM.

Commit:
Use different IPs for records in cornerCasesWildcard zone file

Signed-off-by: Jason Du <xdu@infoblox.com>

chrisohaver reviewed (approved):
LGTM!  IMO an owner of the file plugin should review before merging.

yongtang reviewed (approved):
LGTM
